game/src/layer/elevation.rs

```diff
@@ -132,3 +133,165 @@ impl SteepStreets {
+                        .closest_pt(pt, INTERSECTION_SEARCH_RADIUS)
+                    {
+                        self.tooltip = Some(Text::from(format!("Elevation: {}", elevation)));
+                    }
+                }
+            }
+        }
+
+        Layer::simple_event(ctx, &mut self.panel)
+    }
+    fn draw(&self, g: &mut GfxCtx, app: &App) {
+        self.panel.draw(g);
+        if g.canvas.cam_zoom < app.opts.min_zoom_for_detail {
+            g.redraw(&self.unzoomed);
+        }
+        if let Some(ref txt) = self.tooltip {
+            g.draw_mouse_tooltip(txt.clone());
+        }
+    }
+    fn draw_minimap(&self, g: &mut GfxCtx) {
+        g.redraw(&self.unzoomed);
+    }
+}
+
+impl ElevationContours {
+    pub fn new(ctx: &mut EventCtx, app: &App) -> ElevationContours {
+        let mut low = Distance::ZERO;
+        let mut high = Distance::ZERO;
+        for i in app.primary.map.all_intersections() {
+            low = low.min(i.elevation);
+            high = high.max(i.elevation);
+        }
+
+        let (closest_elevation, unzoomed) = make_contours(ctx, app, low, high);
+
+        let panel = Panel::new(Widget::col(vec![
+            header(ctx, "Elevation"),
+            format!("Elevation from {} to {}", low, high).text_widget(ctx),
+        ]))
+        .aligned_pair(PANEL_PLACEMENT)
+        .build(ctx);
+
+        ElevationContours {
+            tooltip: None,
+            closest_elevation,
+            unzoomed,
+            panel,
+        }
+    }
+}
+
+fn make_contours(
+    ctx: &mut EventCtx,
+    app: &App,
+    low: Distance,
+    high: Distance,
+) -> (FindClosest<Distance>, Drawable) {
+    let bounds = app.primary.map.get_bounds();
+    let mut closest = FindClosest::new(bounds);
+    let mut batch = GeomBatch::new();
+
+    ctx.loading_screen("generate contours", |_, timer| {
+        timer.start("gather input");
+
+        let resolution_m = 30.0;
+        // Elevation in meters
+        let mut grid: Grid<f64> = Grid::new(
+            (bounds.width() / resolution_m).ceil() as usize,
+            (bounds.height() / resolution_m).ceil() as usize,
+            0.0,
+        );
+
+        // Since gaps in the grid mess stuff up, just fill out each grid cell. Explicitly do the
+        // interpolation to the nearest measurement we have.
+        for i in app.primary.map.all_intersections() {
+            // TODO Or maybe even just the center?
+            closest.add(i.elevation, i.polygon.points());
+        }
+        let mut indices = Vec::new();
+        for x in 0..grid.width {
+            for y in 0..grid.height {
+                indices.push((x, y));
+            }
+        }
+        for (idx, elevation) in
+            timer.parallelize("fill out grid", Parallelism::Fastest, indices, |(x, y)| {
+                let pt = Pt2D::new((x as f64) * resolution_m, (y as f64) * resolution_m);
+                let elevation = match closest.closest_pt(pt, INTERSECTION_SEARCH_RADIUS) {
```

Comment:
I kind of wish there was an algorithm that generated contours just from a sparse set of points, but all the crates I've found implement marching squares and need a grid. So even if we don't have any data for any area, we really need to put something there, or we'll see an "island" of 0m. Right now we're storing elevation at each intersection, so this is the simplest strategy of taking the value of the nearest intersection. We could certainly get fancier here -- interpolating between a few of the nearest intersections, using the road's percent grade (which we also store) if we happen to be filling out a cell on part of a road, etc.

Reply:
Also note that our underlying data sources either _provide_ a grid of data, or actually are just contours. So it may seem kind of roundabout to go calculate contours from this. But the reasons are:
1) Generality -- we wouldn't want to use the Seattle contour dataset, because it'd only work here
2) Storage -- we want to keep the map file size small

Reply:
Seems like a reasonable compromise.